pkg/api/short_url.go

```diff
@@ -1,19 +1,128 @@
 package api
 
 import (
+	"bytes"
+	"encoding/json"
 	"fmt"
+	"io"
 	"net/http"
-	"strings"
 
 	"github.com/grafana/grafana/pkg/api/dtos"
 	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/apimachinery/identity"
+	"github.com/grafana/grafana/pkg/services/apiserver/endpoints/request"
 	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
 	"github.com/grafana/grafana/pkg/services/shorturls"
 	"github.com/grafana/grafana/pkg/setting"
 	"github.com/grafana/grafana/pkg/util"
+	"github.com/grafana/grafana/pkg/util/errhttp"
 	"github.com/grafana/grafana/pkg/web"
 )
 
+// r.Post("/api/short-url/"
+func (hs *HTTPServer) getCreateShortURLHandler() web.Handler {
+	if hs.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesShortURLs) {
+		return hs.createKubernetesShortURLsHandler()
+	}
+	return hs.createShortURL
+}
+
+// r.Get("/goto/:uid/"
+func (hs *HTTPServer) getredirectFromShortURLHandler() web.Handler {
+	if hs.Features.IsEnabledGlobally(featuremgmt.FlagKubernetesShortURLs) {
+		return hs.getKubernetesRedirectFromShortURL
+	}
+	return hs.redirectFromShortURL
+}
+
+func (hs *HTTPServer) getKubernetesRedirectFromShortURL(c *contextmodel.ReqContext) {
+	shortURLUID := web.Params(c.Req)[":uid"]
+	if !util.IsValidShortUID(shortURLUID) {
+		return
+	}
+
+	// TODO: Here I need to call k8sStore.Get() (not the service that uses legacy store) to get the short URL resource.
+	shortURL, err := hs.ShortURLService.GetShortURLByUID(c.Req.Context(), c.SignedInUser, shortURLUID)
+	if err != nil {
+		// If we didn't get the URL for whatever reason, we redirect to the
+		// main page, otherwise we get into an endless loops of redirects, as
+		// we would try to redirect again.
+		if shorturls.ErrShortURLNotFound.Is(err) {
+			hs.log.Debug("Not redirecting short URL since not found", "uid", shortURLUID)
+			c.Redirect(hs.Cfg.AppURL, 308)
+			return
+		}
+		hs.log.Error("Short URL redirection error", "err", err)
+		c.Redirect(hs.Cfg.AppURL, 307)
+		return
+	}
+
+	// TODO: Here I need to call k8sStore.Update() (not the service that uses legacy store) to update the short URL last seen value in the resource.
```

Comment:
How could I call the k8s Store from here to update the last seen value?

Reply:
Here's an example of updating through the k8s client: https://github.com/grafana/grafana/blob/main/pkg/api/playlist.go#L284